tests/paramgrill.c

```diff
@@ -340,76 +823,269 @@ static int BMK_seed(winnerInfo_t* winners, const ZSTD_compressionParameters para
+        }
+    }
+}
+
+/* Initialize memotable, immediately mark redundant / obviously infeasible params as */
+static void memoTableInit(U8* memoTable, ZSTD_compressionParameters paramConstraints, const constraint_t target, const varInds_t* varyParams, const int varyLen, const size_t srcSize) {
+    size_t i;
+    size_t arrayLen = memoTableLen(varyParams, varyLen);
+    int cwFixed = !paramConstraints.chainLog || !paramConstraints.windowLog;
+    int scFixed = !paramConstraints.searchLog || !paramConstraints.chainLog;
+    int whFixed = !paramConstraints.windowLog || !paramConstraints.hashLog;
+    int wFixed = !paramConstraints.windowLog;
+    int j = 0;
+    memset(memoTable, 0, arrayLen);
+    cParamZeroMin(&paramConstraints);
+
+    for(i = 0; i < arrayLen; i++) {
+        memoTableIndInv(&paramConstraints, varyParams, varyLen, i);
+        if(ZSTD_estimateCStreamSize_usingCParams(paramConstraints) > (size_t)target.cMem) {
+            memoTable[i] = 255;
+            j++;
+        }
+        if(wFixed && (1ULL << paramConstraints.windowLog) > (srcSize << 2)) {
+            memoTable[i] = 255;
+        }
+        /* nil out parameter sets equivalent to others. */
+        if(cwFixed/* at most least 1 param fixed. */) {
+            if(paramConstraints.strategy == ZSTD_btlazy2 || paramConstraints.strategy == ZSTD_btopt || paramConstraints.strategy == ZSTD_btultra) {
+                if(paramConstraints.chainLog > paramConstraints.windowLog + 1) {
+                    if(memoTable[i] != 255) { j++; }
+                    memoTable[i] = 255;
+                }
+            } else {
+                if(paramConstraints.chainLog > paramConstraints.windowLog) {
+                    if(memoTable[i] != 255) { j++; }
+                    memoTable[i] = 255;
+                }
+            }
+        }
+
+        if(scFixed) {
+            if(paramConstraints.searchLog > paramConstraints.chainLog) {
+                if(memoTable[i] != 255) { j++; }
+                memoTable[i] = 255;
+            }
+        }
+
+        if(whFixed) {
+            if(paramConstraints.hashLog > paramConstraints.windowLog + 1) {
+                if(memoTable[i] != 255) { j++; }
+                memoTable[i] = 255;
+            }
+        }
+    }
+    DEBUGOUTPUT("%d / %d Invalid\n", j, (int)i);
+    if((int)i == j) {
+        DEBUGOUTPUT("!!!Strategy %d totally infeasible\n", (int)paramConstraints.strategy)
+    }
+}
+
+/* frees all allocated memotables */
+static void memoTableFreeAll(U8** mtAll) {
+    int i;
+    if(mtAll == NULL) { return; }
+    for(i = 1; i <= (int)ZSTD_btultra; i++) {
+        free(mtAll[i]);
+    }
+    free(mtAll);
+}
+
+/* inits memotables for all (including mallocs), all strategies */
+/* takes unsanitized varyParams */
+static U8** memoTableInitAll(ZSTD_compressionParameters paramConstraints, constraint_t target, const varInds_t* varyParams, const int varyLen, const size_t srcSize) {
```

Comment:
It's more than an `init`,
this function is actually allocating the space returned,
which means, it will have to be `free()` afterwards.

In our coding convention, we use the word `create` to make such distinction.
Typically `createMemoTable()`,
which implies that a corresponding `freeMemoTable()` necessarily exist.

`initMemoTable()` would be reserved for situations where `U8** mtAll` already exists, 
and needs to be set to some specific values.
This is useful if there is more than one way to reserve `mtAll` area.